src/content_scripts/autocorrect.js

```diff
@@ -0,0 +1,351 @@
+const IS_CHROME = Object.getPrototypeOf(browser) !== Object.prototype;
+
+/**
+ * Get caret position.
+ *
+ * @param {HTMLElement} target
+ * @returns {number|null}
+ */
+function getCaretPosition(target) {
+	// ContentEditable elements
+	if (target.isContentEditable || document.designMode === "on") {
+		target.focus();
+		const _range = document.getSelection().getRangeAt(0);
+		if (!_range.collapsed) {
+			return null;
+		}
+		const range = _range.cloneRange();
+		const temp = document.createTextNode("\0");
+		range.insertNode(temp);
+		const caretposition = target.innerText.indexOf("\0");
+		temp.remove();
+		return caretposition;
+	}
+	// input and textarea fields
+	if (target.selectionStart !== target.selectionEnd) {
+		return null;
+	}
+	return target.selectionStart;
+}
+
+/**
+ * Insert at caret in the given element.
+ * Adapted from: https://www.everythingfrontend.com/posts/insert-text-into-textarea-at-cursor-position.html
+ *
+ * @param {HTMLElement} target
+ * @param {string} atext
+ * @throws {Error} if nothing is selected
+ * @returns {void}
+ */
+function insertAtCaret(target, atext) {
+	// document.execCommand is deprecated, although there is not yet an alternative: https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand
+	// insertReplacementText
+	if (document.execCommand("insertText", false, atext)) {
+		return;
+	}
+
+	// Firefox input and textarea fields: https://bugzilla.mozilla.org/show_bug.cgi?id=1220696
+	if (typeof target.setRangeText === "function") {
+		const start = target.selectionStart;
+		const end = target.selectionEnd;
+
+		if (start != null && end != null) {
+			target.setRangeText(atext);
+
+			target.selectionStart = target.selectionEnd = start + atext.length;
+
+			// Notify any possible listeners of the change
+			const event = document.createEvent("UIEvent");
+			event.initEvent("input", true, false);
+			target.dispatchEvent(event);
+
+			return;
+		}
+	}
+
+	throw new Error("nothing selected");
+}
+
+/**
+ * Insert at caret in the given element and select.
+ *
+ * @param {HTMLElement} target
+ * @param {string} atext
+ * @returns {void}
+ */
+function insertAndSelect(target, atext) {
+	insertAtCaret(target, atext);
+	// ContentEditable elements
+	if (target.isContentEditable || document.designMode === "on") {
+		const range = document.getSelection().getRangeAt(0);
+		range.setStart(range.startContainer, range.startOffset - atext.length);
+	}
+	// input and textarea fields
+	else {
+		target.selectionStart -= atext.length;
+	}
+}
+
+/**
+ * Insert into page.
+ *
+ * @param {string} atext
+ * @returns {void}
+ */
+function insertIntoPage(atext) {
+	return insertAtCaret(document.activeElement, atext);
+}
+
+/**
+ * Count Unicode characters.
+ * Adapted from: https://blog.jonnew.com/posts/poo-dot-length-equals-two
+ * Intl.Segmenter is not yet supported by Firefox/Thunderbird: https://bugzilla.mozilla.org/show_bug.cgi?id=1423593
```

Comment:
Yeah, I wanted to use the new Intl.Segmenter API, but I was not able to since it is not yet supported by Firefox/Thunderbird. I do not believe using it now would provide any advantages, as this add-on is of course not published on the Chrome Web Store nor compatible with MV3 (which is now a requirement for new add-on on the CWS).